### PR TITLE
Add automated MA balance test system

### PR DESCRIPTION
**Pull request by leoCottret**
#### Summary
SUMMARY: Balance "Add automated MA (Martial Art) balance test system"

#### Purpose of change
Original subject: https://github.com/cataclysmbnteam/Cataclysm-BN/pull/1869
image

#### Describe the solution
Add a way to test the balance of MA.
This is basically one big test with options like:
- MA to test
- specific weapon (or weapons from the MA)

#### Describe alternatives you've considered
Using a way to trigger inputs in game through OS inputs (like with autoit for windows).
And extract the damage dealt through the save.
image
Some the basic idea would be:
- Create a world with 0 spawn, disable autosave and the like, spawn walls around player through debug menu, make him naked, equip him, set stats an skills, spawn a monster, hit a few times, kill all monsters through debug menu, heal player (and every X times reequip him), quicksave.
- Extract data from the save
- Repeat
##### Pro
- You can't be closer to the real game than that. No matter how complexe the combat system becomes, you just stimulate a player fighting and get the resulting damage. That's obviously the biggest pro and the real reason to do that.
- There's already a way to spawn monsters, heal players etc. with the debug menu. So it could be done
- CBN is especially adapted to that, everything can be done through keys (mouse input is less reliable), and it seems to get the input very reliably when I tested (each stimulated keys triggered the right action in the game)
- I know AutoIt well enough, everything said above can be done through it
- Easier to debug
- The system can be semi-automatic in a first time if something is hard to implement (eg if selecting a new MA isn't implemented, it could wait for a key input to start, then when you can, you select the MA and press the key input to continue script execution).

##### Cons
- This would work for windows only
- This would be slow, especially if the amount of "player_messages" stored can't be increased
- Extracting data like what weapons can be used with a MA will be done through JSON (might be a pro in some case)
- Even if AutoIt may be the easiest language I learnt, it's still a new syntaxe to learn if someone wants to continue the project in the future

Seeing how hard it is to create a reliable stimulation of a fight taking every systems into account, I'm seriously considering this option.
Otherwise, either there's a way to have a reliable combat system without recreating the game, or we're ok with a partially stimulated one.
(I'm considering doing it anyway to have a way to compare if the cpp version is taking everything into consideration)

#### Testing
- I could extract those data with it, this gives the most damaging weapon with a stats/skills level given, against a specific monster:

| Starting tests for -> STATS: 12 SKILLS: 5 M: mon_zombie_kevlar_2 |                       |         |
|------------------------------------------------------------------|-----------------------|---------|
| Fior Di Battaglia                                                | : sledge hammer       | : 21.5  |
| Barbaran Montante                                                | : sledge hammer       | : 20.6  |
| Ninjutsu                                                         | : loaded stick        | : 15.1  |
| Medieval Swordsmanship                                           | : longsword replica   | : 14.7  |
| Niten Ichi-Ryu                                                   | : nodachi replica     | : 14.4  |
| Bionic Combatives                                                | : monomolecular blade | : 2.6   |
| Starting tests for -> STATS: 15 SKILLS: 7 M: mon_zombie_kevlar_2 |                       |         |
| Barbaran Montante                                                | : mace                | : 34.5  |
| Fior Di Battaglia                                                | : sledge hammer       | : 34.1  |
| Niten Ichi-Ryu                                                   | : bokken              | : 28.2  |
| Ninjutsu                                                         | : loaded stick        | : 26.6  |
| Medieval Swordsmanship                                           | : zweihaender replica | : 25.7  |
| Bionic Combatives                                                | : monomolecular blade | : 4.0   |
| Starting tests for -> STATS: 18 SKILLS: 9 M: mon_zombie_kevlar_2 |                       |         |
| Barbaran Montante                                                | : mace                | : 55.1  |
| Fior Di Battaglia                                                | : sledge hammer       | : 51.8  |
| Niten Ichi-Ryu                                                   | : bokken              | : 48.0  |
| Ninjutsu                                                         | : loaded stick        | : 42.6  |
| Medieval Swordsmanship                                           | : zweihaender replica | : 41.5  |
| Bionic Combatives                                                | : bionic claws        | : 15.5  |
| Starting tests for -> STATS: 12 SKILLS: 5 M: mon_zombie_hulk     |                       |         |
| Bionic Combatives                                                | : monomolecular blade | : 80.9  |
| Fior Di Battaglia                                                | : battle axe          | : 62.4  |
| Ninjutsu                                                         | : monomolecular blade | : 61.5  |
| Barbaran Montante                                                | : battle axe          | : 58.0  |
| Niten Ichi-Ryu                                                   | : hand-forged sword   | : 56.6  |
| Medieval Swordsmanship                                           | : broadsword          | : 49.2  |
| Starting tests for -> STATS: 15 SKILLS: 7 M: mon_zombie_hulk     |                       |         |
| Bionic Combatives                                                | : monomolecular blade | : 101.2 |
| Fior Di Battaglia                                                | : battle axe          | : 79.4  |
| Ninjutsu                                                         | : monomolecular blade | : 74.2  |
| Niten Ichi-Ryu                                                   | : hand-forged sword   | : 73.8  |
| Barbaran Montante                                                | : war hammer          | : 72.3  |
| Medieval Swordsmanship                                           | : broadsword          | : 63.7  |
| Starting tests for -> STATS: 18 SKILLS: 9 M: mon_zombie_hulk     |                       |         |
| Bionic Combatives                                                | : monomolecular blade | : 132.0 |
| Barbaran Montante                                                | : war hammer          | : 91.9  |
| Niten Ichi-Ryu                                                   | : hand-forged sword   | : 91.8  |
| Fior Di Battaglia                                                | : battle axe          | : 91.3  |
| Medieval Swordsmanship                                           | : estoc               | : 80.9  |
| Ninjutsu                                                         | : broadsword          | : 78.7  |
|                                                                  |                       |         |

#### Additional context
- Right now the system needs to trigger too many things manually
- Player/Monster won't hit each other automatically
- Counter attacks needs to be triggered manually
I think this is mainly because I couldn't find a way to advance the turns in the world, but I would like to do it with minimal modifications.
Can it be done?

#### TLDR for concerned people
- It's uncomplete and doesn't take some martial art functionnalities into account, but that's the goal
- Even with a fully done system, it's not meant to bring homogeneity to MA, MA are too complexe to be reduced at a "Damage per Second" comparison only. 
- Eg. A defensive style with low DPS is ok, a style with hgh DPS but no defense is ok, a style with high DPS and defense but needs a rare weapon, CBM + book is ok and the list goes on

### CODE REVIEW

File: tests/melee_test.cpp [hunk 1]
```diff
@@ -237,3 +251,259 @@ TEST_CASE( "Hulk smashing a character", "[.], [melee], [monattack]" )
...
+                    else {
+                        if( p.martial_arts_data->selected_has_weapon( it->get_id() ) ) {
+                            item player_wp( it );
+                            compatible_weapons.push_back( it );
+                        }
+                    }
+                }
+
+                // Reduce number of weapons to "weapons_count_limit" by MA, to speed up tests. Keep the ones with the best DPS
+                // And favor damage type in this order -> Bash, Pierce, Cut
```
Comment (Coolthulhu) [inline tests/melee_test.cpp:353]: Bashing is usually the weakest because of low strength downscaling, which can decrease dmg up to 50% if weapon's bash is way higher than player's str+skill.
Stabbing gets best damage scaling.

`melee.cpp`, `roll_[damage type_]_damage`

File: tests/melee_test.cpp [hunk 2]
```diff
@@ -237,3 +251,259 @@ TEST_CASE( "Hulk smashing a character", "[.], [melee], [monattack]" )
...
+                        g->m.ter_set( p.pos() + point_west, resin_wall );
+                        g->m.ter_set( p.pos() + point_east, resin_wall );
+                        // XXX bottom walls
+                        g->m.ter_set( p.pos() + point_north_west, resin_wall );
+                        g->m.ter_set( p.pos() + point_north, resin_wall );
+                        g->m.ter_set( p.pos() + point_north_east, resin_wall );
+
+
+                        // Wait once for debuff (eg Barbaran montante)
+                        // TODO Verify if it work?
```
Comment (Coolthulhu) [inline tests/melee_test.cpp:448]: If you really want to do it, you could check if the character has all the passive effects from MA and no other effects.

File: tests/melee_test.cpp [hunk 3]
```diff
@@ -237,3 +251,259 @@ TEST_CASE( "Hulk smashing a character", "[.], [melee], [monattack]" )
...
+                            g->u.mod_moves( 300 );
+                            p.process_turn();
+                            z.melee_attack( p );
+                            z.process_turn();
+
+                            // Since counter attacks seems to trigger a much lower rate than in game (meaning close to never), we trigger them manually
+                            if( ( k % 3 == 0 || one_in( 5 ) ) && special_counter->id.str() != "tec_none" ) {
+                                p.melee_attack( z, true, special_counter );
+                            }
+                        }
```
Comment (Coolthulhu) [inline tests/melee_test.cpp:465]: You should probably reset monster's hp and effects every loop. Same for player.
If you want to test dependent effects, you should do it separately and more explicitly. For now, it's safe to disregard "nested" effects.

**Comment by Lamandus** [comment]
is it still worked on?

**Comment by Lamandus** [comment]
Since it is stale for a year, consider reopening it if you are continuing
